### ISSUE TITLE
Library `version` produces files with only the major version on Darwin, but all components on Linux

### ISSUE THREAD

**Issue by e-dant** [mesonbuild/meson#13787]
**Describe the bug**
The [`library`'s](https://mesonbuild.com/Reference-manual_functions.html#library) version field has an explanation beside it:

> A string specifying the version of this shared library, such as 1.1.0. On Linux and OS X, this is used to set the shared library version in the filename, such as libfoo.so.1.1.0 and libfoo.1.1.0.dylib. If this is not specified, soversion is used instead (see above).

On Darwin, the produced shared library seems to instead be `libfoo.1.dylib` on Darwin, and `libfoo.so.1.1.0` on Linux.

**To Reproduce**
https://github.com/e-dant/watcher/blob/release/watcher-c/meson.build#L32

**Expected behavior**
Either updated documentation or the full library version, depending on what was intended.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.) Debian 11 and macOS 15
* what Python version are you using e.g. 3.8.0 -> `Python 3.11.2` (linux), `Python 3.12.6` (macOS)
* what `meson --version` -> 1.5.2
* what `ninja --version` if it's a Ninja build -> 1.11.1 (Linux) 1.12.1 (macOS)

**Comment by e-dant** [comment]
I worry that putting the full version in dylibs by default would break projects, but I'm curious how the documentation ended up the other way.

Was this behavior changed at some point?

Anyway, I think the full version, as an option, would be useful. Maybe even options to install symbolic links around for, say, `liba.so->liba.1.so->liba.1.1.so->liba.1.1.1.so`

**Comment by eli-schwartz** [comment]
On Linux, meson always creates liba.so -> liba.so.1 -> liba.so.1.1.1; you cannot use liba.1.so as this will not be functional in any way shape or form unless you expect end users to link to it using `-la.1`.

It looks like the documentation for macOS is confused between `version: ...` and `darwin_versions: ...`, but I haven't checked the revision history.

**Comment by e-dant** [comment]
Are the symbolic links expected to survive installation in a Python package?

```
$ python3 -m venv virtualenv ; . virtualenv/bin/activate
$ pip install .
...
$ find virtualenv -name '*.so*'
virtualenv/lib/python3.11/site-packages/.wtr_watcher.mesonpy.libs/libwatcher-c.so.0.12.1
```

PS: Sorry, I meant postfix .so.x.y.z on Linux (not .x.y.z.so) earlier

**Comment by eli-schwartz** [comment]
meson-python presumably chooses to only install some of them, because the .whl file format doesn't support symlinks.

The symlinks do not matter for runtime use as you only need the name which is actively bound by the linker, and the libfoo.so without any versioning is used only in development.

(Some linux distros will only install the symlinks for a library, in the "*-dev" package.)

**Comment by smcv** [comment]
> The symlinks do not matter for runtime use

Yes and no. If your library is `libfoo`, ABI version 1.2.3, then the name that is loaded at runtime on Linux (and I believe on most other Unix platforms, except for macOS and OpenBSD) is `libfoo.so.1`, and it's that name that must exist. Normally Meson will make it a symlink `libfoo.so.1 -> libfoo.so.1.2.3`, in which case you need both files; but it's also OK for it to be a regular file `libfoo.so.1`.

You are correct to say that the symlink `libfoo.so -> libfoo.so.1` without any versioning is (normally) only used in development: for example `gcc -lfoo` looks for `libfoo.so`.

> Some linux distros will only install the symlinks for a library, in the "*-dev" package

Distributions that have this split normally put the `libfoo.so` symlink in the `-dev` or `-devel` package, yes. However, if `libfoo.so.1` is a symlink pointing to `libfoo.so.1.2.3`, then both the `libfoo.so.1` symlink and its target usually need to be in the runtime library package.

For example Debian would typically have

```
libfoo-dev
    ${libdir}/libfoo.so -> libfoo.so.1
libfoo1
    ${libdir}/libfoo.so.1 -> libfoo.so.1.2.3
    ${libdir}/libfoo.so.1.2.3 (regular file, not a symlink)
```

**Comment by eli-schwartz** [comment]
Ah thanks, gotcha. I have used multiple distros that don't do a `-dev` split at all. :)
